src/Hosting/SystemdServices/src/ServiceState.cs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Hosting.SystemdServices
+{
+    /// <summary>
+    /// Describes a service state change.
+    /// </summary>
+    public struct ServiceState
+    {
+        private string _state;
```

Comment:
Since this is always going to be ASCII, what about storing the encoded bytes here instead of a `string`? Then there's no need to allocate when writing them to the socket. Unfortunately we can't store a `ReadOnlySpan<byte>` without making this a `ref struct` though, which means we don't get the static data optimization in C#...

Still, the constant values (`Ready`, etc.) could be stored as once-allocated `static readonly byte[]` fields.